internal/telemetry/client.go

```diff
@@ -0,0 +1,407 @@
+// common" variant?
+
+func newmetric(name string, kind metricKind, tags []string, common bool) *metric {
+	return &metric{
+		name:   name,
+		kind:   kind,
+		tags:   append([]string(nil), tags...),
+		common: common,
+	}
+}
+
+func metricKey(name string, tags []string) string {
+	return name + strings.Join(tags, "-")
+}
+
+// Gauge sets the value for a gauge with the given name and tags. If the metric
+// is not language-specific, common should be set to true
+func (c *Client) Gauge(name string, value float64, tags []string, common bool) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if !c.started {
+		return
+	}
+	key := metricKey(name, tags)
+	m, ok := c.metrics[key]
+	if !ok {
+		m = newmetric(name, metricKindGauge, tags, common)
+		c.metrics[key] = m
+	}
+	m.value = value
+	m.ts = float64(time.Now().Unix())
+	c.newMetrics = true
+}
+
+// Count adds the value to a count with the given name and tags. If the metric
+// is not language-specific, common should be set to true
+func (c *Client) Count(name string, value float64, tags []string, common bool) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if !c.started {
+		return
+	}
+	key := metricKey(name, tags)
+	m, ok := c.metrics[key]
+	if !ok {
+		m = newmetric(name, metricKindCount, tags, common)
+		c.metrics[key] = m
+	}
+	m.value += value
+	m.ts = float64(time.Now().Unix())
+	c.newMetrics = true
+}
+
+// Flush sends any outstanding telemetry messages and aggregated metrics to be
+// sent to the backend. Requests are sent in the background
+func (c *Client) Flush() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.flush()
+}
+
+// actual flush implementation, should be called with c.mu locked
+func (c *Client) flush() {
+	submissions := make([]*Request, 0, len(c.requests)+1)
+	if c.newMetrics {
+		c.newMetrics = false
+		r := c.newRequest(RequestTypeGenerateMetrics)
+		payload := &Metrics{
+			Namespace:   c.Namespace,
+			LibLanguage: "golang",
+			LibVersion:  version.Tag,
+		}
+		for _, m := range c.metrics {
+			s := Series{
+				Name:   m.name,
+				Type:   string(m.kind),
+				Tags:   m.tags,
+				Common: m.common,
+			}
+			s.Points = [][2]float64{{m.ts, m.value}}
+			payload.Series = append(payload.Series, s)
+		}
+		r.Payload = payload
+		submissions = append(submissions, r)
+	}
+
+	// copy over requests so we can do the actual submission without holding
+	// the lock. Zero out the old stuff so we don't leak references
+	for i, r := range c.requests {
+		submissions = append(submissions, r)
+		c.requests[i] = nil
+	}
+	c.requests = c.requests[:0]
+
+	go func() {
+		for _, r := range submissions {
+			err := c.submit(r)
+			if err != nil {
+				c.log("telemetry submission failed: %s", err)
+				continue
+			}
+		}
+	}()
+}
+
+var (
+	osName        string
+	osNameOnce    sync.Once
+	osVersion     string
+	osVersionOnce sync.Once
+)
+
+// XXX: is it actually safe to cache osName and osVersion? For example, can the
+// kernel be updated without stopping execution?
```

Comment:
Some distros such as Oracle can do limited [live patching](https://en.wikipedia.org/wiki/Linux_kernel#Live_patching). But I don't know if that ends up bumping the os name/version, nor do I think that we need to be worried about it.